rep-2004.rst

```diff
@@ -0,0 +1,681 @@
+
+For example, packages which provide intra-process communication, inter-process communication, generated message runtime code, node lifecycle, etc. should probably all be considered for 'Level 1'.
+However, a package which provides pose estimation (like ``robot_pose_ekf``\ ) is a generic solution for something that most people need, but is often replaced with a domain specific solution in production, and therefore it should probably not start out as 'Level 1'.
+Such a package may upgrade to 'Level 1' at a later date, if it proves to be a solution that people want to use in their products.
+
+Tools, like ``rostopic``\ , generally do not fall into this category either, but are not categorically excluded.
+For example, it may be the case that the tool which launches and verifies a ROS graph (``ros2launch``\ ) may need to be considered 'Level 1' for use in production systems.
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+Requirements to be considered a 'Level 1' package:
+
+#. **Version Policy:**
+
+   #. Must have a version policy (e.g. ``semver``)
+   #. Must be at a stable version (e.g. for ``semver`` that means version >= 1.0.0)
+   #. Must have a strictly declared public API
+   #. Must have a policy for API stability
+   #. Must have a policy for ABI stability
+   #. Must have a policy that keeps API and ABI stability within a released ROS distribution
+
+#. **Change Control Process:**
+
+   #. Must have all code changes occur through a change request (e.g. pull request, merge request, etc.)
+   #. Must have confirmation of contributor origin (e.g. `DCO  <https://developercertificate.org/>`_, CLA, etc.)
+   #. Must have peer review policy for all change requests (e.g. require one or more reviewers)
+   #. Must have Continuous Integration (CI) policy for all change requests
+   #. Must have documentation policy for all change requests
+
+#. **Documentation:**
+
+   #. Must have documentation for each "feature" (e.g. for ``rclcpp``: create a node, publish a message, spin, etc.)
+   #. Must have documentation for each item in the public API (e.g. functions, classes, etc.)
+   #. Must have a declared license or set of licenses
+   #. Must have a copyright statement in each source file
+   #. Must have a "quality declaration" document, which declares the quality level and justifies how the package meets each of the requirements
+
+      #. *Must have a section in the repository's ``README`` which contains the "quality declaration" or links to it*
+      #. *Must register with a centralized list of 'Level N' packages, if one exists, to allow for peer review of the claim*
+
+#. **Testing:**
+
+   #. Must have system tests which cover all items in the "feature" documentation
+   #. Must have system, integration, and/or unit tests which cover all of the public API
+   #. Code Coverage:
+
+      #. *Must have code coverage tracking for the package*
```

Comment:
Propose to have an acceptance criteria of minimum 90% of code lines covered by unit tests

Reply:
We have explicitly avoided concrete numbers because we don't believe we can give a number that works for every package. The point is for package maintainers to declare what their policy is and describe why they think it is sufficient.

If you want more prescriptive numbers and guidelines, those will be available in the developer guide:

https://github.com/ros2/ros2_documentation/blob/master/source/Contributing/Developer-Guide.rst#code-coverage